cmake/shields.cmake

```diff
@@ -1,18 +1,138 @@
-if(CMAKE_SCRIPT_MODE_FILE AND NOT CMAKE_PARENT_LIST_FILE)
-  # This file was invoked as a script directly with -P:
-  # cmake -P shields.cmake
-  #
-  # Unlike boards.cmake, this takes no OUTPUT_FILE option, but
-  # SHIELD_LIST_SPACE_SEPARATED is required.
-  list(SORT SHIELD_LIST)
-  foreach(shield ${SHIELD_LIST})
-    message("${shield}")
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (c) 2021, Nordic Semiconductor ASA
+
+# Validate shields and setup shields target.
+#
+# This module will validate the SHIELD argument.
+#
+# If a shield implementation is not found for one of the specified shields an
+# error will be raised and list of valid shields will be printed.
+#
+# Outcome:
+# The following variables will be defined when this module completes:
+# - shield_conf_files: List of shield specific Kconfig fragments
+# - shield_dts_files : List of shield specific devicetree files
+# - shield_dts_fixups: List of shield specific devicetree fixups
+#
+# The following targets will be defined when this module completes:
+# - shields: when invoked a list of valid shields will be printed
+#
+# Variable dependencies:
+# - BOARD_ROOT: CMake list of board roots containing board implementations
+#
+# Module dependencies:
+# - extension module.
+
+# Check that SHIELD has not changed.
+zephyr_check_cache(SHIELD WATCH)
+
+if(SHIELD)
+  set(BOARD_MESSAGE "${BOARD_MESSAGE}, Shield(s): ${SHIELD}")
+endif()
+
+if(DEFINED SHIELD)
+  string(REPLACE " " ";" SHIELD_AS_LIST "${SHIELD}")
+endif()
+# SHIELD-NOTFOUND is a real CMake list, from which valid shields can be popped.
+# After processing all shields, only invalid shields will be left in this list.
+set(SHIELD-NOTFOUND ${SHIELD_AS_LIST})
+
+# Use BOARD to search for a '_defconfig' file.
+# e.g. zephyr/boards/arm/96b_carbon_nrf51/96b_carbon_nrf51_defconfig.
+# When found, use that path to infer the ARCH we are building for.
+foreach(root ${BOARD_ROOT})
+  set(shield_dir ${root}/boards/shields)
```

Comment:
> This isn't explicitly documented as being set in the outcome docstring, but I think it is left set in the global scope after the foreach loop exits.

It is, and it should be mentioned.

> Did you consider putting the logic of this module into a function, which uses set(... PARENT_SCOPE) on exactly the variables it means to update in the "outcome" section instead, to avoid ambiguity like this?

Not at this point.
First stage (this PR) is to get the `boilerplate.cmake` reworked into a more reusable state for the use-cases mentioned here: https://github.com/zephyrproject-rtos/zephyr/pull/41301#issue-1083225661
and keeping the current behavior and flow as intact as possible when comparing with today's boilerplate.

When the individual parts are better grouped, next stage will aim at reworking content of the modules themselves, especially when it comes to variables like this, but also to see if code itself can be restructured / modified so some of the small helper module, such as `root.cmake` and `build_configuration.cmake` can be removed, while still ensuring the before referenced use-cases.

I prefer such approach to ensure existing behavior is kept intact for now.

Reply:
> I prefer such approach to ensure existing behavior is kept intact for now.

Sure, that makes good sense as an approach.

>> This isn't explicitly documented as being set in the outcome docstring, but I think it is left set in the global scope after the foreach loop exits.
>
> It is, and it should be mentioned.

.. as long as the docs are kept consistent meanwhile.

Reply:
I have now added:
```
# Variables set by this module and not mentioned above are considered internal
# use only and may be removed, renamed, or re-purposed without prior notice. 
```

Those are existing variables already present in boilerplate.cmake with same scope.
This PR doesn't change those internal variables, their usage, or scope.

So I think having a general clause like this should be sufficient.